pandas/core/frame.py

```diff
@@ -4135,6 +4135,13 @@ def query(self, expr: str, inplace: bool = False, **kwargs):
         kwargs["target"] = None
         res = self.eval(expr, **kwargs)
 
+        if res.ndim == 1:
+            is_bool_result = is_bool_dtype(res)
+        elif res.ndim > 1:
+            is_bool_result = all(is_bool_dtype(x) for x in res.dtypes)
+        if not is_bool_result:
+            msg = f"expr must evaluate to boolean not {res.dtypes}"
+            raise ValueError(msg)
```

Comment:
out of curiosity, as not looked at this in detail, but if we are ensuring a bool indexer, is the following try/except still needed?

Reply:
Indeed it is a common problem with the multidimensional res, so we can share the solution.